File: c++/matmul1.cpp
/************************************************************************\

    matmul1.cpp
    A demo program to read two matrices, performs multiplication, and
    optionally saving it to a text file

    Created: 20170224

\************************************************************************/

#include <iostream>
#include <fstream>
#include <Matrix.hpp>

#include <cstdlib>

//  Performs C = A * B
//  Dimensions:
//    - A(M,K)
//    - B(K,N)
//    - C(M,N)
template <typename _number>
void matrix_multiply(const Matrix<_number> &A,
                     const Matrix<_number> &B,
                     Matrix<_number> &C)
{
    int M = A.numrows();
    int K = A.numcols();
    int N = C.numcols();
    assert(K == B.numrows());
    assert(N == B.numcols());
    int im,ik,in;

    for (im = 0; im < M; ++im)
    {
        for (in = 0; in < N; ++in)
        {
            _number Cmn = 0;
            for (ik = 0; ik < K; ++ik)
                Cmn += A(im,ik) * B(ik,in);
            C(im,in) = Cmn;
        }
    }
}

void matmul1(const char *infile1,
             const char *infile2,
             const char *outfile)
//  read1: Reads a matrix from a file and prints it to stdout
{
    using std::cout;
    using std::cerr;
    using std::endl;

    cout << "## matmul1 ## infile1=" << infile1
         << " infile2=" << infile2 << std::endl;
    std::ifstream inF;

    inF.open(infile1);
    Matrix<double> A(readmatrix1<double>(inF));
    inF.close();
    
    inF.open(infile2);
    Matrix<double> B(readmatrix1<double>(inF));
    inF.close();
    
    //auto A = readmatrix1<double>(inF);
    cout << "A dimensions = " <<  A.numrows() << " " << A.numcols() << endl;
    cout << "B dimensions = " <<  B.numrows() << " " << B.numcols() << endl;

    if (A.numcols() != B.numrows())
    {
        cerr << "Error: mismatched matrix dimensions for multiplication, bailed" << endl;
        abort();
    }

    Matrix<double> C(A.numrows(), B.numcols());

    matrix_multiply(A, B, C);

    if (outfile)
    {
        cout << "   matmul1 ## outfile=" << outfile << endl;
        std::ofstream outF(outfile);
        printmatrix1(C, outF);
    }
    else
    {
        printmatrix1(C, cout);
    }
}


int main(int argc, const char **argv)
{
    if (argc < 3)
    {
        std::cerr << "Needs two input matrix filenames" << std::endl;
        return 1;
    }
    matmul1(argv[1], argv[2], argc > 3 ? argv[3] : NULL);
    return 0;
}

File: python/matmul1.py
#!/usr/bin/env python
#
# Created: 20170227

# Note: In a python program or function, if the first statement
# encountered is a string literal like this, it will be taken as the
# documentation (__doc__) of the program/function.

"""
matmul1.py

Demo program to read two matrices, performs multiplication, and prints
the output to standard output.

This program has some pedagogical value in demonstrating the effect of
hand-crafting loops in python as opposed to using
matrix-multiplication kernel (canned routine).

"""

from __future__ import print_function
import sys
import numpy
from np_helper import loadmatrix1, printmatrix1

def matmul_manual(A, B):
    """Performs manual matrix multiplication with python.
    Avoid doing this in python because of low performance!
    """
    from numpy import asarray, sum
    A = asarray(A)
    B = asarray(B)
    M, K = A.shape
    N = B.shape[1]
    assert A.shape[1] == B.shape[0]
    # Caveat: works only if A & B dtypes are the same:
    C = numpy.zeros((M,N), dtype=A.dtype)
    for i in range(M):
        for j in range(N):
            Cij = 0
            for k in range(K):
                Cij += A[i,k] * B[k,j]
            C[i,j] = Cij
    return C


def matmul_vecdot(A, B):
    """Performs semi-manual matrix multiplication with python,
    using dot product in the innnermost loop.
    """
    from numpy import asarray, sum
    A = asarray(A)
    B = asarray(B)
    M, K = A.shape
    N = B.shape[1]
    assert A.shape[1] == B.shape[0]
    # Caveat: works only if A & B dtypes are the same:
    C = numpy.zeros((M,N), dtype=A.dtype)
    for i in range(M):
        for j in range(N):
            C[i,j] = sum(A[i,:] * B[:,j])
    return C



def matmul_matdot(A, B):
    """Performs matrix multiplication with python using numpy.dot.

    Note that numpy.dot is equivalent to matrix-matrix multiplication
    for 2-D arrays; we explot this fact here!
    """
    from numpy import asarray, sum
    A = asarray(A)
    B = asarray(B)
    M, K = A.shape
    N = B.shape[1]
    assert A.shape[1] == B.shape[0]
    C = numpy.dot(A, B)
    return C



def matmul1(argv):
    if len(argv) < 3:
        print("Needs an input file name on arg1", file=sys.stderr)
        sys.exit(1)

    matfile1 = argv[1]
    matfile2 = argv[2]
    A = loadmatrix1(matfile1)
    B = loadmatrix1(matfile2)
    C = matmul_matdot(A,B)
    printmatrix1(C, float_fmt=" %12.6f")
    #printmatrix1(C)

if __name__ == "__main__":
    matmul1(sys.argv)
